robot/web/stream/server.go

```diff
@@ -319,6 +329,104 @@ func (server *Server) RemoveStream(ctx context.Context, req *streampb.RemoveStre
 	return &streampb.RemoveStreamResponse{}, nil
 }
 
+// GetStreamOptions implements part of the StreamServiceServer. It returns the available dynamic resolutions
+// for a given stream name. The resolutions are scaled down from the original resolution in the camera
+// properties.
+func (server *Server) GetStreamOptions(
+	ctx context.Context,
+	req *streampb.GetStreamOptionsRequest,
+) (*streampb.GetStreamOptionsResponse, error) {
+	if req.Name == "" {
+		return nil, errors.New("stream name is required")
+	}
+	cam, err := camera.FromRobot(server.robot, req.Name)
+	if err != nil {
+		return nil, err
+	}
+	// If the camera properties do not have intrinsic parameters,
+	// we need to sample a frame to get the width and height.
+	var width, height int
+	camProps, err := cam.Properties(ctx)
+	if err != nil || camProps.IntrinsicParams == nil || camProps.IntrinsicParams.Width == 0 || camProps.IntrinsicParams.Height == 0 {
+		server.logger.Debug("width and height not found in camera properties")
+		width, height, err = server.sampleFrameSize(ctx, cam)
+		if err != nil {
+			return nil, fmt.Errorf("failed to sample frame size: %w", err)
+		}
+	} else {
+		width, height = camProps.IntrinsicParams.Width, camProps.IntrinsicParams.Height
+	}
+	scaledResolutions := server.generateResolutions(int32(width), int32(height))
+	resolutions := make([]*streampb.Resolution, 0, 5)
+	for _, res := range scaledResolutions {
+		resolutions = append(resolutions, &streampb.Resolution{
+			Height: res.height,
+			Width:  res.width,
+		})
+	}
+	return &streampb.GetStreamOptionsResponse{
+		Resolutions: resolutions,
+	}, nil
+}
+
+// generateResolutions takes the original width and height of an image and returns
+// a list of the original resolution with 4 smaller width/height options that maintain
+// the same aspect ratio.
+func (server *Server) generateResolutions(width, height int32) []resolution {
```

Comment:
Should we add a check for if the generated resolution gets too small? Maybe break if we get to 1x1? 1x1 is kinda a silly response but idk open to other ideas

Reply:
Yeah I was thinking about this as well. silly input -> silly output.

But, good idea to handle such cases. Added break condition and 2x2 test camera.